[WTF] reorganize `Base64DecodeMode` into flags for more extensibility

dcrousso opened the pull request:
#### a4e626a2b14f7600fd7f65c2aa2d09bb4dd76583
<pre>
[WTF] reorganize `Base64DecodeMode` into flags for more extensibility
<a href="https://bugs.webkit.org/show_bug.cgi?id=275558">https://bugs.webkit.org/show_bug.cgi?id=275558</a>

Reviewed by Darin Adler.

This will assist with implementing `Uint8Array.fromBase64` since that has a couple of other extra behavior modes when decoding base64 content &lt;<a href="https://webkit.org/b/275593">https://webkit.org/b/275593</a>&gt;.

* Source/WTF/wtf/text/Base64.h:
(WTF::base64Decode):
(WTF::base64URLDecode):
* Source/WTF/wtf/text/Base64.cpp:
(WTF::base64DecodeInternal):
(WTF::base64Decode):
(WTF::base64DecodeToString):
Introduce a `Base64DecodeOption` that&apos;s wrapped in an `OptionSet` (for future extensibility) as a parameter for adjusting base64 decoding.

* Source/JavaScriptCore/jsc.cpp:
(functionAtob):
(functionDisassembleBase64):
* Source/WebCore/page/Base64Utilities.cpp:
(WebCore::Base64Utilities::atob):
* Source/WebCore/page/Page.cpp:
(WebCore::Page::userStyleSheetLocationChanged):
* Source/WebCore/platform/graphics/avfoundation/CDMFairPlayStreaming.cpp:
(WebCore::extractSinfData):
* Source/WebCore/platform/network/DataURLDecoder.cpp:
(WebCore::DataURLDecoder::decodeSynchronously):
* Source/WebKit/UIProcess/Inspector/gtk/RemoteWebInspectorUIProxyGtk.cpp:
(WebKit::RemoteWebInspectorUIProxy::platformSave):
* Source/WebKit/UIProcess/Inspector/gtk/WebInspectorUIProxyGtk.cpp:
(WebKit::WebInspectorUIProxy::platformSave):
* Source/WebKit/UIProcess/Inspector/mac/WebInspectorUIProxyMac.mm:
(WebKit::WebInspectorUIProxy::showSavePanel):
* Source/WebKitLegacy/mac/WebCoreSupport/WebInspectorClient.mm:
(WebInspectorFrontendClient::save):
Replace `Base64EncodeMode::DefaultIgnorePadding` with no options (i.e. default behavior).
Replace `Base64EncodeMode::DefaultValidatePadding` with `Base64EncodeOption::ValidatePadding`.
Replace `Base64EncodeMode::DefaultValidatePaddingAndIgnoreWhitespace` with `Base64EncodeOption::ValidatePadding` and `Base64EncodeOption::IgnoreWhitespace`.
Replace `Base64EncodeMode::DefaultIgnoreWhitespaceForQuirk` with `Base64EncodeOption::IgnoreWhitespace`.
Replace `Base64EncodeMode::URL` with `Base64EncodeOption::URL`.

* Tools/TestWebKitAPI/Tests/WTF/Base64.cpp: Added.
(TestWebKitAPI::TEST(Base64, Decode)):
(TestWebKitAPI::TEST(Base64, DecodeValidatePadding)):
(TestWebKitAPI::TEST(Base64, DecodeIgnoreWhitespace)):
(TestWebKitAPI::TEST(Base64, DecodeValidatePaddingIgnoreWhitespace)):
(TestWebKitAPI::TEST(Base64, DecodeURL)):
(TestWebKitAPI::TEST(Base64, DecodeURLValidatePadding)):
(TestWebKitAPI::TEST(Base64, DecodeURLIgnoreWhitespace)):
(TestWebKitAPI::TEST(Base64, DecodeURLValidatePaddingIgnoreWhitespace)):

* Tools/TestWebKitAPI/CMakeLists.txt:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:

Canonical link: <a href="https://commits.webkit.org/280108@main">https://commits.webkit.org/280108@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/99553dbf06f38e089c819667f76cd8c30510ab3c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/55773 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/35096 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/8240 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/58757 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/6204 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/57899 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/42718 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/6402 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/44920 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/4273 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/57802 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/32996 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/48077 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/26054 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/29782 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/5406 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/4347 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/48850 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/51749 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/5675 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/60348 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/55010 "Built successfully and passed tests") | [⏳ 🛠 vision ](https://ews-build.webkit.org/#/builders/visionOS-1-Build-EWS "Waiting in queue, processing has not started yet") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/5804 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/52344 "Passed tests") | 
| | [⏳ 🛠 vision-sim ](https://ews-build.webkit.org/#/builders/visionOS-1-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/48148 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/51841 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/12353 "Built successfully and passed tests") | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-1-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | | [  ~~🛠 jsc-armv7~~](https://ews-build.webkit.org/#/builders/35/builds/76772 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/30927 "Built successfully") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/35/builds/76772 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/32012 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/33093 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/31759 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WTF/wtf/text/Base64.h:
```diff
@@ -71,11 +67,11 @@ String base64EncodeToString(std::span<const uint8_t>, Base64EncodeMode = Base64E
+std::optional<Vector<uint8_t>> base64Decode(std::span<const uint8_t>, OptionSet<Base64DecodeOption> = { });
 
-WTF_EXPORT_PRIVATE String base64DecodeToString(StringView, Base64DecodeMode = Base64DecodeMode::DefaultIgnorePadding);
+WTF_EXPORT_PRIVATE String base64DecodeToString(StringView, OptionSet<Base64DecodeOption> = { });
```
These all used to ignore padding by default. Because we’re changing the default, we now have to touch every call site. We could retain the "ignore padding by default" behavior in one of two ways:
1) Have the flag be `ValidatePadding` instead of `IgnorePadding` with a reverse sense.
2) Have the default value be `{ Base64DecodeOption::IgnorePadding }` instead of `{ }`.

We should do one of those. I prefer (1) because it’s better for URL mode. I don’t think separating this out into distinct flags should require changing the default.

dcrousso replied:
it felt safer to have the default behavior be to validate padding (i.e. one must opt-out of any/all validation steps), but i dont have a strong opinion so i can definitely change it to be `ValidatePadding` instead

darinadler commented on Source/WTF/wtf/text/Base64.h:
```diff
@@ -86,11 +82,6 @@ Vector<uint8_t> base64URLEncodeToVector(std::span<const uint8_t>);
-std::optional<Vector<uint8_t>> base64URLDecode(StringView);
-inline std::optional<Vector<uint8_t>> base64URLDecode(ASCIILiteral literal) { return base64URLDecode(StringView { literal }); }
-std::optional<Vector<uint8_t>> base64URLDecode(std::span<const std::byte>);
-std::optional<Vector<uint8_t>> base64URLDecode(std::span<const uint8_t>);
```
What’s the rationale for removing these?

dcrousso replied:
it felt a bit unnecessary to have these where the only difference was whether `Base64DecodeMode::URL` was provided, but i can put them back if we prefer having the convenience of these utility methods

darinadler reviewed (changes requested):
I understand the idea of breaking this into separate options for more flexibility. But I don’t understand why we’re making all the existing call sites slightly worse, as this patch seems to. I’d like to hear more rationale.

darinadler commented on Source/WTF/wtf/text/Base64.h:
```diff
@@ -86,10 +82,10 @@ Vector<uint8_t> base64URLEncodeToVector(std::span<const uint8_t>);
+std::optional<Vector<uint8_t>> base64URLDecode(StringView, OptionSet<Base64DecodeOption> = { });
+std::optional<Vector<uint8_t>> base64URLDecode(ASCIILiteral literal, OptionSet<Base64DecodeOption> = { });
+std::optional<Vector<uint8_t>> base64URLDecode(std::span<const std::byte>, OptionSet<Base64DecodeOption> = { });
+std::optional<Vector<uint8_t>> base64URLDecode(std::span<const uint8_t>, OptionSet<Base64DecodeOption> = { });
```
I could see leaving the options argument out of these. If you want to pass unusual options, you could just call the base one and pass the URL flag.

darinadler commented on Source/WebCore/platform/network/DataURLDecoder.cpp:
```diff
@@ -156,8 +156,11 @@ static std::optional<Result> decodeSynchronously(DecodeTask& task)
-        auto mode = task.shouldValidatePadding == ShouldValidatePadding::Yes ? Base64DecodeMode::DefaultValidatePaddingAndIgnoreWhitespace : Base64DecodeMode::DefaultIgnoreWhitespaceForQuirk;
-        auto decodedData = base64Decode(PAL::decodeURLEscapeSequences(task.encodedData), mode);
+        OptionSet<Base64DecodeOption> options;
+        options.add(Base64DecodeOption::IgnoreWhitespace);
```
Could just do this in one line.

dcrousso committed:
[WTF] reorganize `Base64DecodeMode` into flags for more extensibility
https://bugs.webkit.org/show_bug.cgi?id=275558

Reviewed by Darin Adler.

This will assist with implementing `Uint8Array.fromBase64` since that has a couple of other extra behavior modes when decoding base64 content <https://webkit.org/b/275593>.

* Source/WTF/wtf/text/Base64.h:
(WTF::base64Decode):
(WTF::base64URLDecode):
* Source/WTF/wtf/text/Base64.cpp:
(WTF::base64DecodeInternal):
(WTF::base64Decode):
(WTF::base64DecodeToString):
Introduce a `Base64DecodeOption` that's wrapped in an `OptionSet` (for future extensibility) as a parameter for adjusting base64 decoding.

* Source/JavaScriptCore/jsc.cpp:
(functionAtob):
(functionDisassembleBase64):
* Source/WebCore/page/Base64Utilities.cpp:
(WebCore::Base64Utilities::atob):
* Source/WebCore/page/Page.cpp:
(WebCore::Page::userStyleSheetLocationChanged):
* Source/WebCore/platform/graphics/avfoundation/CDMFairPlayStreaming.cpp:
(WebCore::extractSinfData):
* Source/WebCore/platform/network/DataURLDecoder.cpp:
(WebCore::DataURLDecoder::decodeSynchronously):
* Source/WebKit/UIProcess/Inspector/gtk/RemoteWebInspectorUIProxyGtk.cpp:
(WebKit::RemoteWebInspectorUIProxy::platformSave):
* Source/WebKit/UIProcess/Inspector/gtk/WebInspectorUIProxyGtk.cpp:
(WebKit::WebInspectorUIProxy::platformSave):
* Source/WebKit/UIProcess/Inspector/mac/WebInspectorUIProxyMac.mm:
(WebKit::WebInspectorUIProxy::showSavePanel):
* Source/WebKitLegacy/mac/WebCoreSupport/WebInspectorClient.mm:
(WebInspectorFrontendClient::save):
Replace `Base64EncodeMode::DefaultIgnorePadding` with no options (i.e. default behavior).
Replace `Base64EncodeMode::DefaultValidatePadding` with `Base64EncodeOption::ValidatePadding`.
Replace `Base64EncodeMode::DefaultValidatePaddingAndIgnoreWhitespace` with `Base64EncodeOption::ValidatePadding` and `Base64EncodeOption::IgnoreWhitespace`.
Replace `Base64EncodeMode::DefaultIgnoreWhitespaceForQuirk` with `Base64EncodeOption::IgnoreWhitespace`.
Replace `Base64EncodeMode::URL` with `Base64EncodeOption::URL`.

* Tools/TestWebKitAPI/Tests/WTF/Base64.cpp: Added.
(TestWebKitAPI::TEST(Base64, Decode)):
(TestWebKitAPI::TEST(Base64, DecodeValidatePadding)):
(TestWebKitAPI::TEST(Base64, DecodeIgnoreWhitespace)):
(TestWebKitAPI::TEST(Base64, DecodeValidatePaddingIgnoreWhitespace)):
(TestWebKitAPI::TEST(Base64, DecodeURL)):
(TestWebKitAPI::TEST(Base64, DecodeURLValidatePadding)):
(TestWebKitAPI::TEST(Base64, DecodeURLIgnoreWhitespace)):
(TestWebKitAPI::TEST(Base64, DecodeURLValidatePaddingIgnoreWhitespace)):

* Tools/TestWebKitAPI/CMakeLists.txt:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:

Canonical link: https://commits.webkit.org/280108@main
